### PR TITLE
Code Quality: Add ThemedIcon Controls and UI Tests

### PR DESCRIPTION
**Pull request by mdtauk**
**Resolved / Related Issues**
This will add the ThemedIcon control and its associated styles and templates, as well as a UITests UserControl to test the control

To prevent extra work, all changes to the Files codebase must link to an approved issue marked as `Ready to build`. Please insert the issue number following the hashtag with the issue number that this Pull Request resolves.
- Closes #

**Steps used to test these changes**

Stability is a top priority for Files and all changes are required to go through testing before being merged into the repo. Please include a list of steps that you used to test this PR.

1. Opened Files ...
2. ...

### CODE REVIEW

File: src/Files.App/Assets/FilesOpenDialog/Files.App.Launcher.exe.sha256
Comment (yaira2) [inline src/Files.App/Assets/FilesOpenDialog/Files.App.Launcher.exe.sha256:1]: Can you revert the changes to the launcher? The easiest way is to copy the files from `main`.

**Comment by mdtauk** [comment]
We need to now wait for @0x5bfa to look at the code, and to add back in the App Service for handling High Contrast with the control, and to add the automated tests they wanted to add.

**Comment by 0x5bfa** [comment]
I think we have to define 'IconType' resource in the ThemeDictionaries in Generic.xaml and create a new dependency property in ThemedIcon.Properties.cs to handle the HC change aware.

Key name should be MandatoryIconType
```xml
<x:String>Any</x:String>
<x:String>Any</x:String>
<x:String>Outline</x:String>

<Setter Property="MandatoryIconType" Value="{ThemeResource MandatoryIconType}" />
```

In this way we can hook HC theme change without our implementation. In fact it's impossible to hook with my service since that service is in Files.App

**Comment by 0x5bfa** [comment]
I think we have to define 'IconType' resource in the ThemeDictionaries in Generic.xaml and create a new dependency property in ThemedIcon.Properties.cs to handle the HC change aware.

Key name should be MandatoryIconType
```xml
<x:String>Any</x:String>
<x:String>Any</x:String>
<x:String>Outline</x:String>

<Setter Property="MandatoryIconType" Value="{ThemeResource MandatoryIconType}" />
```

And ThemedIcon.cs handles the change of that property. If it's 'Any', use the desired type of icon there specified by dep property 'IconType', but if 'Outline' specified use it.

In this way we can hook HC theme change without our implementation. In fact it's impossible to hook with my service since that service is in Files.App

**Comment by 0x5bfa** [comment]
Or Boolean 'IconTypeObliged' in HighContrast theme resource key, it should be true

**Comment by mdtauk** [comment]
> I think we have to define a theme resource in the ThemeDictionaries in Generic.xaml and create a new dependency property in ThemedIcon.Properties.cs to handle the HC change aware.
> 
> Key name should be MandatoryIconType
> 
> ```
> <x:String>Any</x:String>
> <x:String>Any</x:String>
> <x:String>Outline</x:String>
> 
> <Setter Property="MandatoryIconType" Value="{ThemeResource MandatoryIconType}" />
> ```
> 
> And ThemedIcon.cs handles the change of that property. If it's 'Any', use the desired type of icon there specified by dep property 'IconType', but if 'Outline' specified use it.
> 
> In this way we can hook HC theme change without our implementation. In fact it's impossible to hook with my service since that service is in Files.App
---
>Or Boolean 'IconTypeObliged' in HighContrast theme resource key, it should be true

Some questions so I understand @0x5bfa .
- What would be the difference between our IconType and this MandatoryIconType?
- Is this a replacement for the code which bypasses IconType states?
- How would you manually override the Layered and Outline icons, can we set a value separate from the Theme Dictionary.
- We have removed the UseContrast property and visual states, but if we add a boolean which we set in the ThemeDictionary, is that essentially the same thing but used slightly differently?
- If we set this boolean value in the ThemeDictionary, then we don't need the app service, and we can test it at the same time by manually setting the boolean to true in the TestsUI right?

**Comment by 0x5bfa** [comment]
> What would be the difference between our IconType and this MandatoryIconType?

> How would you manually override the Layered and Outline icons, can we set a value separate from the Theme Dictionary.

The point is that we cannot use my service to be aware of HC theme change because that service is in Files.App project and we cannot dereference Files.App (it's called circular reference, which is not allowed). But we want to be aware of it in Files.App.Controls project. In this case, the only thing we can is to make use of ThemeResources(x:Key="HighContrast"). In order to do so, we should have a resource key to determine Windows is HC themed or not.

I change the usage to be easy to understand the concept:

```xml
<x:String x:Key="ThemeModeType">Light</x:String> <!--  in Light key. -->
<x:String x:Key="ThemeModeType">Dark</x:String> <!--  in Dark key. -->
<x:String x:Key="ThemeModeType">HighContrast</x:String> <!--  in HighContrast key. -->
```

In styling in Generic.xaml (create a dependency property 'ThemeModeType in ThemedIcon class beforehand):

```xml
<Setter Property="ThemeModeType" Value="{ThemeResource ThemeModeType}" />
``` 

In ThemedIcon class:

```c#
OnThemeTypeChanged() {} // Change icon type to Outline when it's HighContrast 
```

> Is this a replacement for the code which bypasses IconType states?

Overrides based on theme mode, yes.

> We have removed the UseContrast property and visual states, but if we add a boolean which we set in the ThemeDictionary, is that essentially the same thing but used slightly differently?

Kind of similar. I want the property be hidden but it's not possible.

> If we set this boolean value in the ThemeDictionary, then we don't need the app service, and we can test it at the same time by manually setting the boolean to true in the TestsUI right?

We don't need the service. Yes, we can manually test because the property is disclosed.

File: src/Files.App.Controls/ThemedIcon/ThemedIcon.Consts.cs [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    [TemplateVisualState(Name = NeutralStateName, GroupName = IconColorStateGroupName)]
+    [TemplateVisualState(Name = AccentStateName, GroupName = IconColorStateGroupName)]
+    [TemplateVisualState(Name = ToggleStateName, GroupName = IconColorStateGroupName)]
+    [TemplateVisualState(Name = DisabledColorStateName, GroupName = IconColorStateGroupName)]
+    [TemplateVisualState(Name = DisabledToggleColorStateName, GroupName = IconColorStateGroupName)]
+
+    // Icon IsEnabled Visual States
+    [TemplateVisualState(Name = EnabledStateName, GroupName = EnabledStateGroupName)]
+    [TemplateVisualState(Name = DisabledStateName, GroupName = EnabledStateGroupName)]
+
```
Comment (0x5bfa) [inline src/Files.App.Controls/ThemedIcon/ThemedIcon.Consts.cs:34]:
```suggestion
```

File: src/Files.App.Controls/ThemedIcon/ThemedIcon.Properties.cs [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+
+using Microsoft.UI.Xaml;
+using Microsoft.UI.Xaml.Controls;
+
+namespace Files.App.Controls
+{
+    public partial class ThemedIcon : Control
+    {
+        //
+        //	DEPENDENCY PROPERTY REGISTRATION
```
Comment (0x5bfa) [inline src/Files.App.Controls/ThemedIcon/ThemedIcon.Properties.cs:12]: Remove this kind of comments

File: src/Files.App.Controls/ThemedIcon/ThemedIcon.Properties.cs [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+        //
+
+        // *
+        // Path data string properties
```
Comment (0x5bfa) [inline src/Files.App.Controls/ThemedIcon/ThemedIcon.Properties.cs:16]: As above

File: src/Files.App.Controls/ThemedIcon/ThemedIcon.Properties.cs [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+        /// </summary>
+        public static readonly DependencyProperty OutlineIconDataProperty =
+            DependencyProperty.Register(
+                nameof(OutlineIconData),
+                typeof(string),
+                typeof(ThemedIcon),
+                new PropertyMetadata(string.Empty, (d, e) => ((ThemedIcon)d).OnOutlineIconPropertyChanged((string)e.OldValue, (string)e.NewValue)));
+
+        // *
+        // Enum properties
```
Comment (0x5bfa) [inline src/Files.App.Controls/ThemedIcon/ThemedIcon.Properties.cs:39]: As above

File: src/Files.App.Controls/ThemedIcon/ThemedIcon.Properties.cs [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+        /// </summary>
+        public static readonly DependencyProperty IconColorTypeProperty =
+            DependencyProperty.Register(
+                nameof(IconColorType),
+                typeof(ThemedIconColorType),
+                typeof(ThemedIcon),
+                new PropertyMetadata(ThemedIconColorType.None, (d, e) => ((ThemedIcon)d).OnIconColorTypePropertyChanged((ThemedIconColorType)e.OldValue, (ThemedIconColorType)e.NewValue)));
+
+        // *
+        // Boolean properties
```
Comment (0x5bfa) [inline src/Files.App.Controls/ThemedIcon/ThemedIcon.Properties.cs:62]: As above

File: src/Files.App.Controls/ThemedIcon/ThemedIcon.Properties.cs [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+        /// </summary>
+        public static readonly DependencyProperty IsHighContrastProperty =
+            DependencyProperty.Register(
+                nameof(IsHighContrast),
+                typeof(bool),
+                typeof(ThemedIcon),
+                new PropertyMetadata(defaultValue: false, (d, e) => ((ThemedIcon)d).OnIsHighContrastPropertyChanged((bool)e.OldValue, (bool)e.NewValue)));
+
+        // *
+        // Layers
```
Comment (0x5bfa) [inline src/Files.App.Controls/ThemedIcon/ThemedIcon.Properties.cs:95]: As above

File: src/Files.App.Controls/ThemedIcon/ThemedIcon.Properties.cs [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+                nameof(Layers),
+                typeof(object),
+                typeof(ThemedIcon),
+                new PropertyMetadata(null, (d, e) => ((ThemedIcon)d).OnLayersPropertyChanged((object)e.OldValue, (object)e.NewValue)));
+
+        //
+        //	PUBLIC PROPERTIES
+        //
+
+        // *
```
Comment (0x5bfa) [inline src/Files.App.Controls/ThemedIcon/ThemedIcon.Properties.cs:111]: As above

File: src/Files.App.Controls/ThemedIcon/ThemedIcon.Properties.cs [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+        /// Gets or sets the Outline Icon Path data as a String
+        /// </summary>
+        public string OutlineIconData
+        {
+            get => (string)GetValue(OutlineIconDataProperty);
+            set => SetValue(OutlineIconDataProperty, value);
+        }
+
+        // *
+        // Public enum properties
```
Comment (0x5bfa) [inline src/Files.App.Controls/ThemedIcon/ThemedIcon.Properties.cs:133]: As above

File: src/Files.App.Controls/ThemedIcon/ThemedIcon.Properties.cs [hunk 9]
```diff
@@ -0,0 +1,252 @@
...
+        /// </summary>
+        public ThemedIconColorType IconColorType
+
+        {
+            get => (ThemedIconColorType)GetValue(IconColorTypeProperty);
+            set => SetValue(IconColorTypeProperty, value);
+        }
+
+        // *
+        // Public boolean properties
```
Comment (0x5bfa) [inline src/Files.App.Controls/ThemedIcon/ThemedIcon.Properties.cs:155]: As above

File: src/Files.App.Controls/ThemedIcon/ThemedIcon.Properties.cs [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+        /// Gets or sets a value indicating whether the Icon is in HighContrast state.
+        /// </summary>        
+        public bool IsHighContrast
+        {
+            get => (bool)GetValue(IsHighContrastProperty);
+            set => SetValue(IsHighContrastProperty, value);
+        }
+
+        // *
+        // Public object properties
```
Comment (0x5bfa) [inline src/Files.App.Controls/ThemedIcon/ThemedIcon.Properties.cs:185]: As above

File: src/Files.App.Controls/ThemedIcon/ThemedIcon.Properties.cs [hunk 11]
```diff
@@ -0,0 +1,252 @@
...
+            get => (object)GetValue(LayersProperty);
+            set => SetValue(LayersProperty, value);
+        }
+
+        //
+        //	PROPERTY CHANGE EVENTS
+        //
+
+        //*
+        // Path data string changed events
```
Comment (0x5bfa) [inline src/Files.App.Controls/ThemedIcon/ThemedIcon.Properties.cs:201]: As above

File: src/Files.App.Controls/ThemedIcon/ThemedIcon.Properties.cs [hunk 12]
```diff
@@ -0,0 +1,252 @@
...
+            FilledIconPathUpdate();
+        }
+
+        protected virtual void OnOutlineIconPropertyChanged(string oldValue, string newValue)
+        {
+            OutlineIconPathUpdate();
+        }
+
+        //*
+        // Enum changed events
```
Comment (0x5bfa) [inline src/Files.App.Controls/ThemedIcon/ThemedIcon.Properties.cs:214]: As a bove

File: src/Files.App.Controls/ThemedIcon/ThemedIcon.Properties.cs [hunk 13]
```diff
@@ -0,0 +1,252 @@
...
+            UpdateIconTypeStates();
+        }
+
+        protected virtual void OnIconColorTypePropertyChanged(ThemedIconColorType oldValue, ThemedIconColorType newValue)
+        {
+            UpdateIconColorTypeStates();
+        }
+
+        //*
+        // Boolean changed events
```
Comment (0x5bfa) [inline src/Files.App.Controls/ThemedIcon/ThemedIcon.Properties.cs:227]: As abive

File: src/Files.App.Controls/ThemedIcon/ThemedIcon.Properties.cs [hunk 14]
```diff
@@ -0,0 +1,252 @@
...
+            FilledChanged(newValue);
+        }
+
+        protected virtual void OnIsHighContrastPropertyChanged(bool oldValue, bool newValue)
+        {
+            HighContrastChanged(newValue);
+        }
+
+        //*
+        // Object changed events
```
Comment (0x5bfa) [inline src/Files.App.Controls/ThemedIcon/ThemedIcon.Properties.cs:245]: As above

File: src/Files.App.Controls/ThemedIcon/ThemedIcon.cs [hunk 15]
```diff
@@ -0,0 +1,362 @@
...
+        private void ThemeSettings_OnHighContrastChanged(object sender, bool e)
+        {
+            HighContrastChanged(e);
+        }
+
+        private void HighContrastChanged(bool value)
+        {
+            // handles HighContrast property change
+
+            _isHighContrast = value;
```
[the remark this answers is not in the document]
Comment (mdtauk) [reply src/Files.App.Controls/ThemedIcon/ThemedIcon.cs:236]: I kept it consistent to my handling of the other boolean properties, to separate Explicit values set on the ThemedIcon control, and those we handle in code, or in response to event hooks

File: src/Files.App.Controls/ThemedIcon/ThemedIconLayer/ThemedIconLayer.Properties.cs [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+// Licensed under the MIT License. See the LICENSE.
+
+using Microsoft.UI.Xaml;
+
+namespace Files.App.Controls
+{
+    public partial class ThemedIconLayer
+    {
+        //
+        // Dependency Property Registration
```
[the remark this answers is not in the document]
Comment (mdtauk) [reply src/Files.App.Controls/ThemedIcon/ThemedIconLayer/ThemedIconLayer.Properties.cs:11]: The header is there to help group the types of properties

File: src/Files.App.Controls/ThemedIcon/ThemedIconLayer/ThemedIconLayer.cs [hunk 17]
```diff
@@ -0,0 +1,94 @@
...
+        protected override void OnApplyTemplate()
+        {
+            // Initialize with default values
+            base.OnApplyTemplate();
+
+            UpdateIconLayerState();
+            LayerPathDataChanged(PathData);
+        }
+
+        void LayerTypeChanged(ThemedIconLayerType layerType) => UpdateIconLayerState();
```
[the remark this answers is not in the document]
Comment (mdtauk) [reply src/Files.App.Controls/ThemedIcon/ThemedIconLayer/ThemedIconLayer.cs:28]: This was something c# cleanup did, I do prefer the full brackets, so I will do that

File: src/Files.App.Controls/ThemedIcon/ThemedIconLayer/ThemedIconLayer.Properties.cs (hunk 16 above)
[the remark this answers is not in the document]
Comment (0x5bfa) [reply src/Files.App.Controls/ThemedIcon/ThemedIconLayer/ThemedIconLayer.Properties.cs:11]: Actually, I have a plan to use C# attributes.

```c#
// ThemedIcon.Properties.cs
[DependencyProperty(Name = "IconType", Type = typeof(ThemedIconType), Fallback = OnIconTypeChanged)]
public partial class ThemedIcon
{
}
``` 

That complex code turns into single code!

[the remark this answers is not in the document]
Comment (0x5bfa) [reply src/Files.App.Controls/ThemedIcon/ThemedIconLayer/ThemedIconLayer.Properties.cs:11]: But if want to keep it's ok since the control implementation won't be changed so frequently

File: src/Files.App.Controls/ThemedIcon/ThemedIcon.cs (hunk 15 above)
[the remark this answers is not in the document]
Comment (0x5bfa) [reply src/Files.App.Controls/ThemedIcon/ThemedIcon.cs:236]: I don't think that is necessary

File: src/Files.App.Controls/ThemedIcon/ThemedIcon.Properties.cs (hunk 2 above)
Comment (mdtauk) [reply src/Files.App.Controls/ThemedIcon/ThemedIcon.Properties.cs]: Replaced with Regions

File: src/Files.App.Controls/ThemedIcon/ThemedIconLayer/ThemedIconLayer.Properties.cs (hunk 16 above)
[the remark this answers is not in the document]
Comment (0x5bfa) [reply src/Files.App.Controls/ThemedIcon/ThemedIconLayer/ThemedIconLayer.Properties.cs]: Source Generator catches that attributes and parses it.
- https://github.com/files-community/Files/blob/main/src/Files.Core.SourceGenerator/DependencyPropertyGenerator.cs
- https://github.com/files-community/Files/blob/main/src/Files.App/Data/Attributes/DependencyPropertyAttribute.cs
- https://github.com/files-community/Files/blob/main/src/Files.App/UserControls/PathBreadcrumb.xaml.cs#L10

**Comment by 0x5bfa** [comment]
Yair and I hates region directive.
We should keep out of using it

**Comment by 0x5bfa** [comment]
Can you format all xaml by right clicking the ThemedIcon folder and clicking Format XAML.

**Comment by mdtauk** [comment]
> Yair and I hates region directive. We should keep out of using it

So comments are better?  When everything is only separated with a single blank line, its hard to glance and read for me.

**Comment by mdtauk** [comment]
> Can you format all xaml by right clicking the ThemedIcon folder and clicking Format XAML.

Done

File: src/Files.App.Controls/ThemedIcon/ThemedIconLayer/ThemedIconLayer.Properties.cs (hunk 16 above)
[the remark this answers is not in the document]
Comment (mdtauk) [reply src/Files.App.Controls/ThemedIcon/ThemedIconLayer/ThemedIconLayer.Properties.cs]: Are we able to move or copy those to the Controls project, so ThemedIcon and future controls can use them?

As long as functionality is not hampered, and it works well - it will simplify the code.

**Comment by mdtauk** [comment]
I have added a Custom color option.  You set the IconColorType to Custom, and then set the Color property to a Brush (This allows more than solid colours)
image
image
image

**Review by 0x5bfa — APPROVED** [review]
I'm astonished by the quality of this control ThemedIcon 😲
I think this control should be the default of WinUI (this icon control looks like SF Symbol font from macOS from perspective of usability and customizability).

Requesting review from @yaira2 

What I should denote is:
- We'll create UI Tests for other controls (not UserControls). Thus, we should have Frame and NavigationView.
- This UITests project can do unit tests but this project should not test that but we should create a new one (Files.App.UnitTests). Yet still there's problem that this kind of WinUI test project can't reference Files.App project
- You see some region/endregion directives. Those are only in Properties partial class to hold dependency properties, LOC of which tend to be large enough. I'm planning to use DepenendecyPropertyAttribute to make them simple:
  ```c#
  [DependencyProperty(Name = "IconType", IsNullable = false, PropertyChanged = "OnIconTypeChanged", IsSetterPrivate = false)]
  ```
  Besides, it looks like that Partial Property is in a preview of C# 13. We can expect this in the next version of C# and easily adopt soon.

File: src/Files.App.Controls/ThemedIcon/Styles/Icons.Action.xaml [hunk 18]
```diff
@@ -0,0 +1,223 @@
+<!--  Copyright (c) 2024 Files Community. Licensed under the MIT License. See the LICENSE.  -->
+<ResourceDictionary
+	xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
+	xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
+	xmlns:local="using:Files.App.Controls">
+
+	<Style
+		x:Key="App.ThemedIcons.Action_Copying"
```
Comment (yaira2) [inline src/Files.App.Controls/ThemedIcon/Styles/Icons.Action.xaml:8]: Can `Actions` be removed or alternatively, the underscore be changed to a period?

**Comment by mdtauk** [comment]
> I'm astonished by the quality of this control ThemedIcon 😲
> 
> I think this control should be the default of WinUI (this icon control looks like SF Symbol font from macOS from perspective of usability and customizability).
> 

Nice of you to say that, it's not close to the complexity and richness of SF Symbol, but it should be easy enough to work with for the app.

File: src/Files.App.Controls/ThemedIcon/Styles/Icons.Action.xaml (hunk 18 above)
Comment (mdtauk) [reply src/Files.App.Controls/ThemedIcon/Styles/Icons.Action.xaml]: Done

**Review by yaira2 — APPROVED** [review]
LGTM

**Comment by mdtauk** [comment]
I have noticed an issue with the Icon when resizing.  The Layered Icon does not update as the others do.  It's not a show stopper, as we will use the icons at a static size, but it is something I would like to fix.


https://github.com/user-attachments/assets/f4dce036-a850-4123-8261-5d429235ad08

**Comment by mdtauk** [comment]
Layered Icon now resizes properly with both Filled and Outline

https://github.com/user-attachments/assets/77a7e94f-513b-4a84-a37b-f7dc4183fb34

File: src/Files.App.Controls/ThemedIcon/ThemedIcon.cs [hunk 19]
```diff
@@ -0,0 +1,387 @@
...
+            // Runs the ToggleChanged event, to set initial value, if the ToggleButton's isChecked is true
+            //
+            // Check if owner Control is an AppBarToggleButton
+            // Hooks onto Event handlers when IsChecked and IsUnchecked runs
+            // Runs the ToggleChanged event, to set initial value, if the AppBarToggleButton's isChecked is true
+            //
+            // Gets the owner Control
+            // Hooks onto Event handlers when IsEnabledChanged runs
+            // Runs the EnabledChanged event to set initial value
+            //
```
[the remark this answers is not in the document]
Comment (mdtauk) [reply src/Files.App.Controls/ThemedIcon/ThemedIcon.cs:144]: I thought it may be useful if others come to maintain the control in the future.  It also helped me keep track of the flow as I was writing the code.

I am still a beginner with C#

File: src/Files.App.Controls/ThemedIcon/ThemedIconLayer/ThemedIconLayer.cs [hunk 20]
```diff
@@ -0,0 +1,118 @@
...
+            // Code to take the PathData string, and convert it to an actual path
+
+            // If there is no Path Data, we just exit
+            // If there is, we create a new Geometry and insert our Path Data string
+            // We check the Template Part exists, and is the right type, and assign it's data
+
+            if (string.IsNullOrEmpty(pathData))
+                return;
+
+            var geometry = (Geometry)XamlReader.Load(
```
[the remark this answers is not in the document]
Comment (mdtauk) [reply src/Files.App.Controls/ThemedIcon/ThemedIconLayer/ThemedIconLayer.cs:106]: We are converting a string to path data, as strings are shareable where as geometry is not.

File: tests/Files.App.UITests/App.xaml [hunk 21]
```diff
@@ -8,9 +8,25 @@
...
+				</ResourceDictionary>
+				<ResourceDictionary x:Key="Light">
+					<SolidColorBrush x:Key="PageBackground" Color="{ThemeResource LayerOnMicaBaseAltFillColorDefault}" />
+					<StaticResource x:Key="ToggleBackground" ResourceKey="AccentFillColorDefaultBrush" />
+					<SolidColorBrush x:Key="ToggleDisabledBackground" Color="{ThemeResource AccentFillColorDisabled}" />
+				</ResourceDictionary>
+				<ResourceDictionary x:Key="HighContrast">
+					<SolidColorBrush x:Key="PageBackground" Color="{ThemeResource SystemColorWindowColor}" />
+					<SolidColorBrush x:Key="ToggleBackground" Color="{ThemeResource SystemColorHighlightColor}" />
+					<SolidColorBrush x:Key="ToggleDisabledBackground" Color="{ThemeResource SystemColorGrayTextColor}" />
```
[the remark this answers is not in the document]
Comment (mdtauk) [reply tests/Files.App.UITests/App.xaml:27]: These are in the TestUI so the visual examples respond to theme changes to maintain legibility.

image
image
image

File: src/Files.App.Controls/ThemedIcon/ThemedIcon.cs [hunk 22]
```diff
@@ -0,0 +1,387 @@
...
+            // Handles changes to the IconType and setting the correct Visual States.
+
+            // Handles the two IconType states, based on the ThemedIcon.IconType value
+            // as well as states derived from owner controls, and other properties
+
+            // We first check for isToggled and isFilled icon types and states
+            // Then we check for Contrast and Disabled states, to replace Layered with Outline and set EnabledStates
+            // Finally we assigned Filled and Layered states, and default otherwise to Outline
+
+            if (_isToggled is true || IsToggled is true || _isFilled is true || IsFilled is true)
```
[the remark this answers is not in the document]
Comment (mdtauk) [reply src/Files.App.Controls/ThemedIcon/ThemedIcon.cs:297]: The control can be explicitly set to IsToggled, but also we respond to some owner control's toggled state - so the control's property, overrides the owner's state.

File: src/Files.App.Controls/ThemedIcon/ThemedIcon.cs [hunk 23]
```diff
@@ -0,0 +1,387 @@
...
+            }
+            else
+            {
+                VisualStateManager.GoToState(
+                    this,
+                    IconType switch
+                    {
+                        ThemedIconTypes.Layered => LayeredTypeStateName,
+                        _ => OutlineTypeStateName,
+                    },
```
[the remark this answers is not in the document]
Comment (mdtauk) [reply src/Files.App.Controls/ThemedIcon/ThemedIcon.cs:323]: We originally had 3 states handled with this code, I removed the Filled ThemedIconType, and made it a Bool property.  I simply removed one of the lines, and left the code as is.  @0x5bfa re-wrote the nested if statements I wrote originally.

File: src/Files.App.Controls/ThemedIcon/ThemedIconLayer/ThemedIconLayers.cs [hunk 24]
```diff
@@ -0,0 +1,14 @@
...
+// Licensed under the MIT License. See the LICENSE.
+
+using System.Collections.Generic;
+
+namespace Files.App.Controls
+{
+    /// <summary>
+    /// A collection of Layers for the ThemedIcon's Layered IconType
+    /// </summary>
+    public sealed class ThemedIconLayers : List<ThemedIconLayer>
```
[the remark this answers is not in the document]
Comment (mdtauk) [reply src/Files.App.Controls/ThemedIcon/ThemedIconLayer/ThemedIconLayers.cs:11]: @0x5bfa Will need to answer that :)

File: src/Files.App.Controls/ThemedIcon/ThemedIconColorType.cs [hunk 25]
```diff
@@ -0,0 +1,48 @@
...
+{
+    /// <summary>
+    /// Defines IconColorTypes for <see cref="ThemedIcon"/>.
+    /// </summary>
+    public enum ThemedIconColorType
+    {
+        None,
+
+        /// <summary>
+        /// Icon color type of <see cref="ThemedIcon"/> is Normal. Default Value.
```
[the remark this answers is not in the document]
Comment (mdtauk) [reply src/Files.App.Controls/ThemedIcon/ThemedIconColorType.cs:15]: I will add some more comments.

File: src/Files.App.Controls/ThemedIcon/ThemedIcon.Consts.cs [hunk 26]
```diff
@@ -0,0 +1,84 @@
+﻿// Copyright (c) 2024 Files Community
+// Licensed under the MIT License. See the LICENSE.
+
+using Microsoft.UI.Xaml;
+using Microsoft.UI.Xaml.Controls;
+
+namespace Files.App.Controls
+{
+    #region Template Parts
```
[the remark this answers is not in the document]
Comment (0x5bfa) [reply src/Files.App.Controls/ThemedIcon/ThemedIcon.Consts.cs:9]: We guide not to use in our coding guidelines.
I thought I deleted, @mdtauk could you replace them with single comment like before?

File: src/Files.App.Controls/ThemedIcon/ThemedIconLayer/ThemedIconLayers.cs (hunk 24 above)
[the remark this answers is not in the document]
Comment (0x5bfa) [reply src/Files.App.Controls/ThemedIcon/ThemedIconLayer/ThemedIconLayers.cs:11]: This is used in re-styles in ThemedIcon.Git.xaml for example to hold layers.
I'm wondering if we can use something else

```xml
<!-- <Style x:Name="App.ThemedIcon.Git.Push" /> -->
<ThemedIconLayers>
  <ThemedIconLayer />
  <ThemedIconLayer />
  <ThemedIconLayer />
<ThemedIconLayers/>
```

File: src/Files.App.Controls/ThemedIcon/ThemedIcon.cs (hunk 23 above)
[the remark this answers is not in the document]
Comment (mdtauk) [reply src/Files.App.Controls/ThemedIcon/ThemedIcon.cs:325]: @0x5bfa If you can tell me what the Ternary version of this code should be, I can do it.  As it stands, I don't understand how I would do it lol

File: src/Files.App.Controls/ThemedIcon/ThemedIcon.cs [hunk 27]
```diff
@@ -0,0 +1,387 @@
...
+        private void UpdateIconContent()
+        {
+            // Updates PathData and Layers
+
+            FilledIconPathUpdate();
+            OutlineIconPathUpdate();
+            LayeredIconContentUpdate();
+        }
+
+        private void FilledIconPathUpdate()
```
[the remark this answers is not in the document]
Comment (mdtauk) [reply src/Files.App.Controls/ThemedIcon/ThemedIcon.cs:63]: Fixed

File: src/Files.App.Controls/ThemedIcon/ThemedIconLayer/ThemedIconLayer.cs [hunk 28]
```diff
@@ -0,0 +1,118 @@
...
+                    LayerType switch
+                    {
+                        ThemedIconLayerType.Alt => AltStateName,
+                        _ => BaseStateName,
+                    },
+                    true);
+            }
+        }
+  
+        void SetPathData(string pathData, FrameworkElement element)
```
[the remark this answers is not in the document]
Comment (mdtauk) [reply src/Files.App.Controls/ThemedIcon/ThemedIconLayer/ThemedIconLayer.cs]: fixed

File: src/Files.App.Controls/ThemedIcon/ThemedIcon.Consts.cs (hunk 26 above)
[the remark this answers is not in the document]
Comment (mdtauk) [reply src/Files.App.Controls/ThemedIcon/ThemedIcon.Consts.cs]: Done

File: src/Files.App.Controls/ThemedIcon/ThemedIcon.cs (hunk 19 above)
[the remark this answers is not in the document]
Comment (mdtauk) [reply src/Files.App.Controls/ThemedIcon/ThemedIcon.cs:143]: I have put the longer comments into collapsable comment blocks

File: src/Files.App.Controls/ThemedIcon/ThemedIcon.cs (hunk 23 above)
[the remark this answers is not in the document]
Comment (0x5bfa) [reply src/Files.App.Controls/ThemedIcon/ThemedIcon.cs:315]: IconType is ThemedIconTypes.Layered ? LayeredTypeStateName : OutlineTypeStateName


```cs
if (condition)
  return consequence;
else
  return alternative;
```

to

```cs 
condition ? consequence : alternative;
```

This can be used not only in return state but everywhere in any statement.

File: src/Files.App.Controls/ThemedIcon/ThemedIcon.cs [hunk 29]
```diff
@@ -0,0 +1,387 @@
...
+            }
+
+            ownerControl = this.FindAscendant<Control>();
+
+            if (ownerControl != null)
+            {
+                ownerControl.IsEnabledChanged += OwnerControl_IsEnabledChanged;
+
+                EnabledChanged(ownerControl.IsEnabled);
+            }
```
[the remark this answers is not in the document]
Comment (marcelwgn) [reply src/Files.App.Controls/ThemedIcon/ThemedIcon.cs:173]: Ah I see, got it.

**Comment by 0x5bfa** [comment]
I updated IsHighContrast's setter to be internal, so classes in Files.App can't be change this property value.

File: src/Files.App.Controls/ThemedIcon/ThemedIconLayer/ThemedIconLayers.cs (hunk 24 above)
[the remark this answers is not in the document]
Comment (0x5bfa) [reply src/Files.App.Controls/ThemedIcon/ThemedIconLayer/ThemedIconLayers.cs:11]: I don't come up any other ideas to hold a list of objects in xaml.
If @marcelwgn has, I'd love to hear.

[the remark this answers is not in the document]
Comment (marcelwgn) [reply src/Files.App.Controls/ThemedIcon/ThemedIconLayer/ThemedIconLayers.cs:11]: It seems there is no good way right now and it still an open feature request: https://github.com/microsoft/microsoft-ui-xaml/issues/2232

I think its the best solution we have right now.

**Review by yaira2 — APPROVED** [review]
LGTM
